### DISCUSSION TITLE
Containerization

### DISCUSSION THREAD

**Discussion by AstraLuma** [Palats/mapshot#26] [General]
Because I have a terminal case of devops, I'm trying to encapsulate mapshot, factorio, and all the necessary dependencies into a container. I've attached where I'm at.

[Dockerfile.txt](https://github.com/Palats/mapshot/files/9792658/Dockerfile.txt)

Unfortunately, factorio just seems to hang? Verbose log attached.

`$ docker run --rm -v $PWD/saves:/opt/factorio/saves:ro -v $PWD/output:/opt/factorio/script-output mapshot mapshot --logtostderr --factorio_verbose render _autosave1`

```
   4.040 Crop bitmaps.
   4.042 Parallel Sprite Loader initialized (threads: 11)
   4.101 Initial atlas bitmap size is 16384
   4.103 Created atlas bitmap 2048x1136 [icon, not-compressed, mipmap, linear-minification, linear-magnification, linear-mip-level]
   4.103 Created virtual atlas pages 4096x4096x0
   4.858 Texture processor created (2048). GPU accelerated compression Supported: yes, Enabled: yes/yes. Test passed. YCoCgDXT PSNR: 35.83, BC3 PSNR: 33.82
   4.893 Parallel Sprite Loader initialized (threads: 11)
   5.018 Sprites loaded
   5.096 Generated mipmaps (5) for atlas [0] of size 2048x1136   
   5.096 Custom mipmaps uploaded.
   5.109 Warning GraphicsInterfaceOpenGL.cpp:245: SDL_GL_SetSwapInterval failed: That operation is not supported
   5.109 Factorio initialised
   5.128 Warning GraphicsInterfaceOpenGL.cpp:245: SDL_GL_SetSwapInterval failed: That operation is not supported
```

[mapshot.log](https://github.com/Palats/mapshot/files/9792672/mapshot.log)

Is this where you need VirtualGL? The only symptom I have is "there's no log output, not seeing files, but factorio's running."

**Comment by AstraLuma** [comment]
Also, if anyone has done the work to get `xvfb-run` to forward signals, let me know.

**Comment by AstraLuma** [comment]
xvfb manpage mentioned xwud, found out factorio is asking about a mod conflict.

**Reply by AstraLuma** [reply]
Starting to wonder if I should use xvnc instead of xvfb.

**Comment by AstraLuma** [comment]
Cool, I got it to generate a map.

This signal forwarding is still a problem if you ever want to stop the process, and you should absolutely pull the mods & map from a running game--if the game prompts for any reason, the process just hangs.

**Reply by Palats** [reply]
For the signal forwarding - signal management is currently [super naive](https://github.com/Palats/mapshot/blob/ea9007a937f9003f530e7e95c1617af3895444cd/cmd/render.go#L163) (aka, non existent beyond default). I don't remember the mechanism of signal propagation to the child, but it tends to be full of traps, so that probably could be improved (I imagine that's to react to things like SIGINT?).

"pull the mods & map from a running game" - I'm not sure what you mean here?

**Reply by AstraLuma** [reply]
I mean, if the game or mod versions are different from what the map last ran with, or out of sync with each other, it can trigger a dialog box and halt the mapshot process in a headless context.

**Reply by AstraLuma** [reply]
and I think it's currently `xvfb-run` that's not forwarding signals.

**Comment by AstraLuma** [comment]
Still working on this, trying to figure out how to make Factorio/SDL happy and actually render things.

**Reply by AstraLuma** [reply]
turns out the executor i was using (Cirrus CLI) ignores the entrypoint.

**Comment by changchiyou** [comment]
Any news?

**Reply by AstraLuma** [reply]
It ran locally. As long as nothing made a game-stopping prompt.

Use in the context of a pipeline system is more complicated, since a bunch of them mess with the container in some way.
